### PR TITLE
Review of Taras' UBNT technical challenge

### PR DESCRIPTION
**Pull request by janperner-ubnt**
Taras, 

Please see my review comments in-line.
I'd like to see your reaction. Please use GitHub review feature for conversation.
Please do not merge this PR, so we can delete it, once we're done, in order to keep the conversation private.

Thanks.

### CODE REVIEW

File: src/Services/Server.hxx [hunk 1]
```diff
@@ -33,7 +36,7 @@ public:
      * Blocking method that waits until a client connects to the server
      * @return new file descriptor
      */
-    int Accept();
+    int Accept(); // could return an object, e.g. Session, that would close a socket in its dtor
```
Comment (petrytar) [inline src/Services/Server.hxx]: agree

File: src/Services/Server.hxx [hunk 2]
```diff
@@ -62,16 +65,20 @@ private:
     void OnError(const char * errMsg) const;
 
     // server address
-    sockaddr_in servAddr_;
+    sockaddr_in servAddr_; // why hold as a member?
```
Comment (petrytar) [inline src/Services/Server.hxx]: I was not sure that they would not be useful in the future. Should be removed

File: src/Services/Server.hxx [hunk 3]
```diff
@@ -62,16 +65,20 @@ private:
...
 
     // client address
-    sockaddr_in cliAddr_;
+    sockaddr_in cliAddr_; // why hold as a member?
```
Comment (petrytar) [inline src/Services/Server.hxx]: see comment above

File: src/Services/Server.hxx [hunk 4]
```diff
@@ -62,16 +65,20 @@ private:
...
+    // Shall be wrapped in a single purpose object, e.g. in SocketHandle, also reused by a Client, that would accept a socket handle in its ctor,
+    // and call close() in its dtor().
+    // That object will be hold by value by Server and Client objects, will delete copy ctor and copy assignment operator (moves would be enabled)
+    // so there would be no need to define dtor in Server and Client classes, which implies implementing rule-of-five.
     int accSocketFD_;
 
     // client socket file descriptor
-    int cliSocketFD_;
+    int cliSocketFD_; // this should not be a part of the socket class, it's overriden always if new client connects
 };
```
Comment (petrytar) [inline src/Services/Server.hxx]: Ok, thanks

File: src/Services/SystemCalculator.cxx [hunk 5]
```diff
@@ -27,11 +27,12 @@ std::string SystemCalculator::GetCPU() const
     std::string result = "";
 
     // execute script get_cpu_usage.sh with time 0.1s
-    std::unique_ptr<FileHandle> fHandle(new FileHandle(GET_CPU_USAGE_SCRIPT));
+    std::unique_ptr<FileHandle> fHandle(new FileHandle(GET_CPU_USAGE_SCRIPT)); // why hold on a heap??
```
Comment (petrytar) [inline src/Services/SystemCalculator.cxx]: no valuable arguments

File: src/Services/SystemCalculator.hxx [hunk 6]
```diff
@@ -25,7 +25,7 @@ public:
      * Returns CPU usage by reading from /proc/stat (exec scripts/get_cpu_usage.sh script)
      * @return percent CPU usage by reading from /proc/stat
      */
-    std::string GetCPU() const;
+    std::string GetCPU() const; // any reason for making them instance methods, they should be static
```
Comment (petrytar) [inline src/Services/SystemCalculator.hxx]: agree

File: src/Services/SystemCalculator.hxx [hunk 7]
```diff
@@ -68,9 +68,9 @@ private:
             }
         }
 
-  private:
-    FILE* f_;
-};
+    private:
+        FILE* f_; // what about std::unique_ptr<FILE, decltype(pclose)> ?
```
Comment (petrytar) [inline src/Services/SystemCalculator.hxx]: proposed solution is awesome! Thanks!!
should be replaced to std::unique_ptr<FILE, decltype(&pclose)> f_(popen(GET_CPU_USAGE_SCRIPT, "r"), pclose);

**Comment by petrytar** [comment]
Thank you for good and fast review, very useful

File: src/Daemon.cxx [hunk 8]
```diff
@@ -17,8 +17,10 @@ void OnError(const char *msg)
 // Process client command (e.g. cpu, mem)
 std::string ProcessClientCommand(const char * command)
 {
+    // making them static?
     const char * cpuCommand = "cpu";
     const char * memCommand = "mem";
+    // why std::shared_ptr here?
```
[the remark this answers is not in the document]
Comment (janperner-ubnt) [reply src/Daemon.cxx]: Yeah, that's what I see, but the questions are:

- why dynamically (on a heap)?
- and why using the shared_ptr?

File: src/Services/Client.cxx [hunk 9]
```diff
@@ -19,13 +19,15 @@ Client::Client(const char * hostname, int portno)
...
     servAddr_.sin_family = AF_INET;
-    memcpy((char *)&servAddr_.sin_addr.s_addr,
+    memcpy((char *)&servAddr_.sin_addr.s_addr, // C-style casts here
            (char *)server_->h_addr,
            server_->h_length);
     servAddr_.sin_port = htons(portno);
 
     if (connect(socketFD_,(struct sockaddr *) &servAddr_, sizeof(servAddr_)) < 0)
         OnError("ERROR connecting");
+
```
[the remark this answers is not in the document]
Comment (janperner-ubnt) [reply src/Services/Client.cxx]: Ok, got it.

File: src/Services/Server.hxx [hunk 10]
```diff
@@ -9,6 +9,7 @@
 namespace ubi
 {
 /* A simple server in the internet domain using TCP. The port number is passed as an argument */
+// not decorated by final, so virtual dtor shall be defined; or class shall be decorated by final
```
[the remark this answers is not in the document]
Comment (janperner-ubnt) [reply src/Services/Server.hxx]: absolutely

File: src/Services/Server.hxx [hunk 11]
```diff
@@ -19,6 +20,8 @@ public:
      */
     Server(int portno);
 
+    // copy ctor/copy assignment operator special members should be forbidden
+
```
[the remark this answers is not in the document]
Comment (janperner-ubnt) [reply src/Services/Server.hxx]: no need for:
```
Server() = delete;
```
The compiler doesn't generate the default `ctor` when there exists a user-defined `ctor`.

**Pull request closed without merging**